backend/src/test/kotlin/app/ehrenamtskarte/backend/auth/webservice/ApiTokenServiceTest.kt

```diff
@@ -0,0 +1,165 @@
+            assertEquals(numberOfTokensBefore + 1, numberOfTokensAfter)
+        }
+    }
+
+    @Test
+    fun createApiToken_throwsForbiddenExceptionWhenNotAuthorized() {
+        every { mockJwtPayload.adminId } returns TestAdministrators.NUERNBERG_PROJECT_STORE_MANAGER.id
+
+        transaction {
+            assertThrows(ForbiddenException::class.java) {
+                ApiTokenService().createApiToken(1, mockDfe)
+            }
+            assertFalse(
+                Authorizer.mayAddApiTokensInProject(
+                    AdministratorEntity.find { Administrators.id eq mockJwtPayload.adminId }.single()
+                )
+            )
+        }
+    }
+
+    @Test
+    fun createApiToken_throwsExceptionWhenAdminNotFound() {
+        every { mockJwtPayload.adminId } returns 1234
+
+        transaction {
+            assertThrows(ForbiddenException::class.java) {
+                ApiTokenService().createApiToken(1, mockDfe)
+            }
+        }
+    }
+
+    @Test
+    fun deleteApiToken_deletesTokenSuccessfully() {
+        every { mockJwtPayload.adminId } returns TestAdministrators.KOBLENZ_PROJECT_ADMIN.id
+
+        transaction {
+            ApiTokens.insert {
+                it[ApiTokens.id] = 1
+                it[projectId] = 1
+                it[creatorId] = TestAdministrators.KOBLENZ_PROJECT_ADMIN.id
+                it[tokenHash] = ByteArray(1)
+                it[expirationDate] = LocalDate.now()
+            }
+
+            val tokenExists = ApiTokens.select { ApiTokens.id eq 1 }.count() > 0
+            assertTrue(tokenExists)
+
+            ApiTokenService().deleteApiToken(1, mockDfe)
+
+            val tokenNoLongerExists = ApiTokens.select { ApiTokens.id eq 1 }.singleOrNull() != null
+            assertTrue(tokenNoLongerExists)
+        }
+    }
+
+    @Test
+    fun deleteApiToken_deletesFromOtherAdminTokenSuccessfully() {
+        every { mockJwtPayload.adminId } returns TestAdministrators.KOBLENZ_PROJECT_ADMIN.id
+
+        transaction {
+            ApiTokens.insert {
+                it[ApiTokens.id] = 1
+                it[projectId] = 1
+                it[creatorId] = TestAdministrators.NUERNBERG_PROJECT_ADMIN.id
+                it[tokenHash] = ByteArray(1)
+                it[expirationDate] = LocalDate.now()
+            }
+
+            val tokenExists = ApiTokens.select { ApiTokens.id eq 1 }.count() > 0
+            assertTrue(tokenExists)
+
+            ApiTokenService().deleteApiToken(1, mockDfe)
+
+            val tokenNoLongerExists = ApiTokens.select { ApiTokens.id eq 1 }.singleOrNull() != null
+            assertTrue(tokenNoLongerExists)
+        }
+    }
+
+    @Test
+    fun deleteApiToken_deletesNoTokenFormOtherProjekt() {
+        every { mockJwtPayload.adminId } returns TestAdministrators.KOBLENZ_PROJECT_ADMIN.id
+
+        transaction {
+            val numberOfTokensBefore = ApiTokens.selectAll().count()
+            ApiTokenService().deleteApiToken(3, mockDfe)
```

Comment:
shouldn't here also be an exception thrown if an admin from another project tries to delete a token.
is there even a token with the `id=3` ?

Reply:
You are right, there is no token with id 1, i adjusted it. With the current implementation there is no exception thrown for this case, it is just handled as "this token does not exist", as in the database request we filter for project matching. I fixed the test to test what it is expected to do.